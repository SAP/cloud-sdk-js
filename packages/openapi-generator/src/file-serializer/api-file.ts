import { codeBlock, documentationBlock, unixEOL } from '@sap-cloud-sdk/util';
import {
  serializeImports,
  Import
} from '@sap-cloud-sdk/generator-common/internal';
import {
  OpenApiApi,
  OpenApiOperation,
  OpenApiReferenceSchema
} from '../openapi-types';
import { collectRefs, getUniqueRefs } from '../schema-util';
import { serializeOperation } from './operation';

/**
 * Serialize an API representation to a string representing the resulting API file.
 * @param api - Representation of an API.
 * @param serviceName - Service name for which the API is created.
 * @returns The serialized API file contents.
 * @internal
 */
export function apiFile(api: OpenApiApi, serviceName: string): string {
  const imports = serializeImports(getImports(api));
  const apiDoc = apiDocumentation(api, serviceName);
  const apiContent = codeBlock`
export const ${api.name} = {
  ${api.operations.map(operation => serializeOperation(operation)).join(',\n')}
};
`;

  return [imports, apiDoc, apiContent].join(unixEOL);
}

/**
 * Get the unique reference schemas for all request body types in the given operation list.
 * @param operations - The given operation list.
 * @returns The list of unique referenced body type schemas.
 */
function collectRefsFromOperations(
  operations: OpenApiOperation[]
): OpenApiReferenceSchema[] {
  return getUniqueRefs(
    operations.reduce(
      (referenceTypes, operation) => [
        ...referenceTypes,
        ...collectRefs(
          ...[
            operation.requestBody?.schema,
            operation.response,
            ...operation.queryParameters.map(({ schema }) => schema)
          ]
        )
      ],
      []
    )
  );
}

function getImports(api: OpenApiApi): Import[] {
  const refs = collectRefsFromOperations(api.operations).map(
    requestBodyType => requestBodyType.schemaName
  );

  const refImports = {
    names: refs,
    moduleIdentifier: './schema',
    typeOnly: true
  };
  const openApiImports = {
    names: ['OpenApiRequestBuilder'],
    moduleIdentifier: '@sap-cloud-sdk/openapi'
  };

  return [openApiImports, refImports];
}

/**
 * @internal
 */
export function apiDocumentation(api: OpenApiApi, serviceName: string): string {
  return documentationBlock`
  Representation of the '${api.name}'.
  This API is part of the '${serviceName}' service.`;
}
