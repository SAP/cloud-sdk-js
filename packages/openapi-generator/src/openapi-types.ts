import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  operationName: string;
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

const supportedMethods = {
  get: 'get',
  put: 'put',
  post: 'post',
  patch: 'patch',
  delete: 'delete',
  head: 'head',
  options: 'options'
} as const;

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export type Method = typeof supportedMethods[keyof typeof supportedMethods];

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get supported methods.
 * @returns Methods supported by OpenApi and SAP Cloud SDK.
 */
export function methods(): Method[] {
  return Object.values(supportedMethods);
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  parameterName: string;
  parameterType: string;
}
