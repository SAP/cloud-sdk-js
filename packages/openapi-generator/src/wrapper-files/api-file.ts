import { codeBlock, partition, unique } from '@sap-cloud-sdk/util';
import { OpenApiDocument, OpenApiOperation } from '../openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get the content of the SAP Cloud SDK API wrapper.
 * @param openApiDocument Parsed service.
 * @returns The generated code for the SDK API wrapper.
 */
export function apiFile(openApiDocument: OpenApiDocument): string {
  const requestBodyTypes = getRequestBodyTypes(openApiDocument);
  return codeBlock`
import { OpenApiRequestBuilder } from '@sap-cloud-sdk/core';
import { DefaultApi } from './openapi/api';
${
  requestBodyTypes
    ? `import { ${requestBodyTypes} } from './openapi/model';`
    : ''
}

export const ${openApiDocument.apiName} = {
  ${getOperations(openApiDocument)}
};
`;
}

/**
 * Get the types for all request body types in the given service.
 * @param openApiDocument Parsed service metadata.
 * @returns The list of body types as a string.
 */
function getRequestBodyTypes(openApiDocument: OpenApiDocument): string {
  const bodyTypes = openApiDocument.operations
    .map(operation => operation.requestBody?.parameterType)
    .filter(requestBody => typeof requestBody !== 'undefined') as string[];

  return unique(bodyTypes).join(', ');
}

/**
 * Get all operation representations for the given service.
 * @param openApiDocument Parsed service.
 * @returns All operations as a string.
 */
function getOperations(openApiDocument: OpenApiDocument): string {
  return openApiDocument.operations
    .map(operation => getOperation(operation))
    .join(',\n');
}

/**
 * Get the string representation of one operation.
 * @param operation Operation to serialize.
 * @returns The operation as a string.
 */
function getOperation(operation: OpenApiOperation): string {
  const params = getParams(operation);
  const argsQuestionMark = params.every(param => !param.required) ? '?' : '';
  const paramsArg = params.length
    ? codeBlock`args${argsQuestionMark}: {
  ${params
    .map(param => `${param.name}${param.required ? '' : '?'}: ${param.type}`)
    .join(',\n')}
}`
    : '';
  const requestBuilderParams = [
    'DefaultApi',
    `'${operation.operationName}'`,
    ...params.map(param => `args${argsQuestionMark}.${param.name}`)
  ];

  return codeBlock`
${
  operation.operationName
}: (${paramsArg}) => new OpenApiRequestBuilder<DefaultApi, '${
    operation.operationName
  }'>(
  ${requestBuilderParams.join(',\n')}
)`;
}

interface Parameter {
  type: string;
  name: string;
  required?: boolean;
}

function getParams(operation: OpenApiOperation): Parameter[] {
  const parameters = [
    ...operation.parameters,
    ...getRequestBodyParams(operation)
  ];

  const [required, optional] = partition(parameters, param => !!param.required);
  return [...required, ...optional];
}

function getRequestBodyParams(operation: OpenApiOperation): Parameter[] {
  return operation.requestBody
    ? [
        {
          name: operation.requestBody.parameterName,
          type: operation.requestBody.parameterType,
          required: operation.requestBody.required
        }
      ]
    : [];
}
