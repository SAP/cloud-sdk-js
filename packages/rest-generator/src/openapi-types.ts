import { $Refs } from '@apidevtools/swagger-parser';
import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  serviceDirName: string;
  refs: $Refs;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export enum Method {
  get = 'get',
  put = 'put',
  post = 'post',
  patch = 'patch',
  delete = 'delete',
  head = 'head',
  options = 'options'
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  name: string;
  type: string;
}
