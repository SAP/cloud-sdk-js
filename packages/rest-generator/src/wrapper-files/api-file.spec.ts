import { OpenApiDocument, OpenApiRequestBody } from '../openapi-types';
import { apiFile } from './api-file';

describe('api-file', () => {
  it('creates api file content for operations with parameters and no request bodies', () => {
    const openApiDocument: OpenApiDocument = {
      apiName: 'TestServiceApi',
      operations: [
        {
          operationName: 'getEntity',
          method: 'get',
          parameters: [
            {
              in: 'path',
              name: 'id',
              type: 'string',
              required: true
            },
            {
              in: 'query',
              name: 'limit',
              type: 'number'
            }
          ],
          pattern: 'test/{id}'
        },
        {
          operationName: 'deleteEntity',
          method: 'delete',
          parameters: [
            {
              in: 'path',
              name: 'id',
              type: 'string',
              required: true
            }
          ],
          pattern: 'test/{id}'
        }
      ],
      serviceDirName: 'test-service'
    };
    expect(apiFile(openApiDocument)).toMatchSnapshot();
  });

  it('creates api file content for operation with request body', () => {
    const openApiDocument: OpenApiDocument = {
      apiName: 'TestServiceApi',
      operations: [
        {
          operationName: 'createEntity',
          method: 'post',
          parameters: [],
          requestBody: {
            parameterName: 'body',
            parameterType: 'Body',
            required: true
          } as OpenApiRequestBody,
          pattern: 'test'
        },
        {
          operationName: 'updateEntity',
          method: 'patch',
          parameters: [
            {
              in: 'path',
              name: 'id',
              type: 'string',
              required: true
            }
          ],
          requestBody: {
            parameterName: 'body',
            parameterType: 'Body',
            required: true
          } as OpenApiRequestBody,
          pattern: 'test/{id}'
        }
      ],
      serviceDirName: 'test-service'
    };
    expect(apiFile(openApiDocument)).toMatchSnapshot();
  });

  it('creates api file content for operation with no parameters or request body', () => {
    const openApiDocument: OpenApiDocument = {
      apiName: 'TestServiceApi',
      operations: [
        {
          operationName: 'getEntity',
          method: 'get',
          parameters: [],
          pattern: 'test'
        }
      ],
      serviceDirName: 'test-service'
    };
    expect(apiFile(openApiDocument)).toMatchSnapshot();
  });
});
