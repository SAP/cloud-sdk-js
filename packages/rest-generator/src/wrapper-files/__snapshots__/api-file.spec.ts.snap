// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-file creates api file content for operation with no parameters or request body 1`] = `
"import { RestRequestBuilder } from '@sap-cloud-sdk/core';
import { DefaultApi } from './openapi/api';


export const TestServiceApi = {
  getEntity: () => new RestRequestBuilder<DefaultApi, 'getEntity'>(
    DefaultApi,
    'getEntity'
  )
};"
`;

exports[`api-file creates api file content for operation with request body 1`] = `
"import { RestRequestBuilder } from '@sap-cloud-sdk/core';
import { DefaultApi } from './openapi/api';
import { Body } from './openapi/model';

export const TestServiceApi = {
  createEntity: (body: Body) => new RestRequestBuilder<DefaultApi, 'createEntity'>(
    DefaultApi,
    'createEntity',
    body
  ),
  updateEntity: (id: string, body: Body) => new RestRequestBuilder<DefaultApi, 'updateEntity'>(
    DefaultApi,
    'updateEntity',
    id,
    body
  )
};"
`;

exports[`api-file creates api file content for operations with parameters and no request bodies 1`] = `
"import { RestRequestBuilder } from '@sap-cloud-sdk/core';
import { DefaultApi } from './openapi/api';


export const TestServiceApi = {
  getEntity: (id: string, limit?: number) => new RestRequestBuilder<DefaultApi, 'getEntity'>(
    DefaultApi,
    'getEntity',
    id,
    limit
  ),
  deleteEntity: (id: string) => new RestRequestBuilder<DefaultApi, 'deleteEntity'>(
    DefaultApi,
    'deleteEntity',
    id
  )
};"
`;
